src/main/webapp/script.js

```diff
@@ -451,26 +456,68 @@ function createRecipeElement(recipe, pictureWrap) {
   const timeElement = clone.querySelector(".recipe-card-time");
   timeElement.innerHTML = "Preparation Time: " + recipe["readyInMinutes"] + " minutes";
   
-  const tagElements = clone.querySelector(".recipe-card-tags");
-  createRecipeCardTags(recipe['id'], tagElements);
+  const tagElements = clone.querySelector(".container");
+  const carouselId = clone.querySelector("#myCarousel");
+  carouselId.id = "myCarousel" + recipe['id'];
+
+  const carouselLinks = clone.querySelector(".carousel-indicators");
+  console.log(carouselLinks.className);
+  const carousel = clone.querySelector(".carousel-inner");
+  createRecipeCardTags(recipe['id'], carouselLinks, carousel);
+
+  const carouselLeft = clone.querySelector(".left");
+  const carouselRight = clone.querySelector(".right");
+  carouselLeft.href = "#myCarousel" + recipe['id'];
+  carouselRight.href = "#myCarousel" + recipe['id'];
 
   const tagTextElement = clone.querySelector(".tag-input");
+  const label = clone.querySelector(".form-label");
+  tagTextElement.addEventListener('click', () => {
+    label.style.display = "none";
+  });
 
   const addTagElement = clone.querySelector(".add-tag-button");
   addTagElement.addEventListener('click', () => {
-    const newTagName = (tagTextElement.value).trim();
-    if (newTagName != "") {
-      const params = new URLSearchParams();
-      params.append('tag-name', newTagName);
-      params.append('recipe-id', recipe['id']);
-
-      fetch('/tag', {method: 'POST', body: params}).then(response => response.json()).then((tagList) => {
-        tagElements.innerHTML = "";
-        createRecipeCardTags(recipe['id'], tagElements);
-        postSavedRecipe(recipe);
-        refreshTagNameSelection();
-      });
-    }
+    fetch('/login').then(response => response.json()).then((userInfo) => {
+      if (!userInfo.isLoggedIn) {
+        const loginPopup = document.createElement('div');
+        loginPopup.className = "login-full-page";
+
+        loginPopup.onclick = function() {
+          loginPopup.style.display = "none";
+        }
+
+        const loginPopupText = document.createElement('h6');
+        loginPopupText.className = "login-full-text";
+        loginPopupText.innerText = "Please login to add tags and view nutrition information";
+        
+        const loginLink = document.createElement('a');
+        loginLink.className = "login-full-link upload";
+        loginLink.innerText = "LOGIN";
+        loginLink.href = userInfo.loginUrl;
+
+        document.body.append(loginPopup);
+        loginPopup.append(loginPopupText);
+        loginPopup.append(loginLink);
+        document.body.style.overflow = "hidden";
+        document.getElementById("display-recipes").style.opacity = "0.2";
```

Comment:
Consider extracting these into a helper function for not-logged-in users.

Also extract a helper function for the else-branch